[FEAT] 환율 API 연동 및 DB 저장 기능 구현 (#16)

Jaejeong-Lee opened the pull request:
## 📌 관련 이슈번호
<!-- 관련있는 이슈 번호(#000)을 적어주세요.
  해당 pull request merge와 함께 이슈를 닫으려면 
  closed #Issue_number(이곳에 이슈 번호 작성)를 적어주세요 -->
* Closed #16 

## 📌 해결하는 데 얼마나 걸렸나요?  (예상 작업 시간 / 실제 작업 시간)
<!-- ISSUE에 작성한 시간 대비 실제 작업시간을 적어가며, 객관적인 개발 예상시간을 그려보는 눈을 길러 봅시다. -->
3일/2일

## 📌 PR Point
<!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요. 변경 사항 및 이유 작성 -->
// 환율 API 연동 및 DB 저장 기능
- 수출입은행 현재환율 api 를 활용하여 환율 정보 불러옴
- 살 때 환율과 팔 때 환율은 api에 존재하지 않으므로, 매매기준율에서 임의로 수정해서 사용 예정.(feat. 멘토님 피드백)
- 현재 구동을 확인하기 위해 post 방식으로 호출하는 함수를 controller에 선언해둠
- {{baseUrl}}/exchange-rate/post 로 해당 기능 호출 가능.

// (중요) 변경 사항
- rootConfig 파일의 @ComponentScan 어노테이션에 "org.scoula.config" -> "org.scoula"로 변경

// Database.sql 문 변경 - DB 테이블 구조 일부 변경
- exchange_rate 테이블 구조 변경
- currencyName 추가 (ex. 미국 달러, 대한민국 원)
- 환율 변수 타입 int -> double로 수정

## 📌 참고 사항
<!-- 참고할 사항이 있다면 적어주세요. -->

<img width="1617" height="694" alt="스크린샷 2025-08-04 오후 2 47 05" src="https://github.com/user-attachments/assets/65b99c1c-ed5d-467e-925e-134788721c56" />

Jaejeong-Lee committed:
[feat] 환율 api 연동 및 DB 저장 기능 구현
- database.sql 쿼리문 수정 (exchange_rate 테이블)
- rootConfig 파일 componentScan 부분 수정
- 현재 Post 요청으로 환율 저장 함수 실행하도록 처리.

Jaejeong-Lee committed:
[chore]  필요없는 주석 삭제 및 스케쥴러 코드 추가

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
```
p1 : 우리는 swagger로 테스트 할거라 상단에
`@Api(tags = "Exchange")` 추가 부탁드려요

Jaejeong-Lee replied:
수정했습니다!

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,24 @@
+@RequestMapping("/exchange-rate")
+public class ExchangeController {
+
+    // 생성자 주입
```
p1 : 이런 주석들은 당연한거라 삭제해주세여

Jaejeong-Lee replied:
넵!

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,24 @@
+        service.fetchAndSaveExchangeRates();
+        return ResponseEntity.ok().build();
+    }
+}
```
p1 : 스웨거 관리 , 성공 응답 통일화가 하나도 되어 있지 않아요 .. 지금은 처음 사용하는 거니까 아래코드 대로 작성 해주시고 다음번엔 이 형태 살려서 작성해주세요 

### step 1 
common > exception > enums > SuccessCode안에 아래의 문구 추가
`	SAVE_EXCHAGE_RATE_SUCCESS(HttpStatus.OK, "환율 정보 저장 성공입니다."),` 

### step2 
Controller 수정 
```
@ApiOperation(value = "[JWT] 환율 정보 저장", notes = "환율 정보 저장 API")
	@ApiResponses(value = {
		@ApiResponse(code = 200, message = "환율 정보 저장 성공", response = SuccessResponse.class),
		@ApiResponse(code = 404, message = "해당 유저가 존재하지 않습니다.", response = ErrorResponse.class)
	})
	@PostMapping()
	public SuccessNonResponse saveExchangerates() {
                 service.fetchAndSaveExchangeRates();
		return SuccessNonResponse.success(SuccessCode.SAVE_EXCHAGE_RATE_SUCCESS);
	}

```

Jaejeong-Lee replied:
넵 반영했습니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/dto/ExchangeRateApiDTO.java:
p1 : 지금 Dto가 class로 되어 있는데 아마 record에도  `@JsonIgnoreProperties(ignoreUnknown = true) // 필요없는 json 데이터 제외
`, `    @JsonProperty("cur_unit")
` 이런 어노테이션들 적용가능할텐데 변경하면 좋을거 같아요

Jaejeong-Lee replied:
class 대신 record를 최대한 써보겠습니다. 근데 환율 정보를 받을 때에는 부득이하게 class를 써야만 하는 경우가 생기네요. record를 사용하면 값을 저장할 때 setter 대신 인자를 넣는 방식으로 객체를 사용해야하는데, 변수가 많다보니 너무 길어지고, class를 사용하는 게 맞는 거 같아 class를 사용하고자 합니다. 다른 경우에서는 최대한 record를 써볼게요!

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/dto/ExchangeRateDTO.java:
p1 : 지금 Dto가 class로 되어 있는데 아마 record로 변경 부탁드려요

Jaejeong-Lee replied:
최대한 바꾸겠습니다.

hellozo0 replied:
@Jaejeong-Lee 이부분 Buidler라도 써서 set하는게 좋을거 같아요 
근데 사실 set으로 하드 코딩하는거 보다 record 써서 매핑 해주는게 코드적으로 더 간편하고 길이도 더 짧아진다고 생각하는데 한번 더 고민해보시면 좋을거 같아요  지금의 set 방식은 좋지 않다고 생각합니다. 

## 1. record 쓸경우 
```
public record ExchangeRateDTO(
    String currencyCode,
    String currencyName,
    Double baseExchangeRate,
    Double rateBuy,
    Double rateSell,
    LocalDateTime exchangeRateDate,
    LocalDateTime createdAt,
    LocalDateTime updatedAt
) {}
```

```
for (ExchangeRateApiDTO dto : apiDtoList) {
    String raw = dto.getBaseExchangeRate();
    Double baseRate = Double.parseDouble(raw.replace(",", ""));
    LocalDateTime now = LocalDateTime.now();

    ExchangeRateDTO entity = new ExchangeRateDTO(
        dto.getCurrencyCode(),
        dto.getCurrencyName(),
        baseRate,
        baseRate, // rateBuy
        baseRate, // rateSell
        exchangeRateDate,
        now,
        now
    );

    apiMapper.saveDataToVO(entity);
}
```

## 2. class 사용 
```
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ExchangeRateDTO {
    private String currencyCode;
    private String currencyName;
    private Double baseExchangeRate;
    private Double rateBuy;
    private Double rateSell;
    private LocalDateTime exchangeRateDate;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
}
```
```
for (ExchangeRateApiDTO dto : apiDtoList) {
    String raw = dto.getBaseExchangeRate();
    Double baseRate = Double.parseDouble(raw.replace(",", ""));
    LocalDateTime now = LocalDateTime.now();

    ExchangeRateDTO entity = ExchangeRateDTO.builder()
        .currencyCode(dto.getCurrencyCode())
        .currencyName(dto.getCurrencyName())
        .baseExchangeRate(baseRate)
        .rateBuy(baseRate)
        .rateSell(baseRate)
        .exchangeRateDate(exchangeRateDate)
        .createdAt(now)
        .updatedAt(now)
        .build();

    apiMapper.saveDataToVO(entity);
}
```

hellozo0 replied:
@Jaejeong-Lee  그리고 mapping 작업은 record 안에 자체 매핑 함 수 만들 수 있습니다. 아래 예시입니다~ 
굳이 데이터 매핑작업을 Service에서 보여줄 필요는 없을거 같아요
 
 ```
public record AirTicketDto(
	Long airlineId,
	String reservationCode,
	String departureDate,
	AirportInfo departure,
	AirportInfo arrival,
	String flightNumber,
	String baggage,
	String seatClass
) {
	public static AirTicketDto from(AirTicketVO vo) {
		return new AirTicketDto(
			vo.getAirlineId(),
			vo.getReservationCode(),
			vo.getDepartureDate(),
			new AirportInfo(vo.getDepartureCity(), vo.getDepartureAirport(), vo.getDepartureTime()),
			new AirportInfo(vo.getArrivalCity(), vo.getArrivalAirport(), vo.getArrivalTime()),
			vo.getFlightNumber(),
			vo.getBaggage(),
			vo.getSeatClass().getValue()
		);
	}
}
```

hellozo0 commented on trippy-server/src/main/java/org/scoula/domain/exchange/ExchangeRateVO.java:
p1 : 지금 Dto가 class로 되어 있는데 아마 record로 변경 부탁드려요

hellozo0 commented on trippy-server/src/main/java/org/scoula/mapper/exchange/ExchangeRateApiMapper.java:
```diff
@@ -0,0 +1,9 @@
+
+@Mapper
+public interface ExchangeRateApiMapper {
+    void insertDataToVO(ExchangeRateDTO exchangeRateDTO);
```
p1 : 주로 백엔드에서는 저장할때는 insert 보다는 save를 더 많이 사용해요. 
그리고 save만 해도 해당 DB에 저장하겠다는 의미가 충분해서 변경하면 좋을거 같아요

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateService.java:
p1 : 저희는 Service interface 안쓰고 Service 바로 써요, 그래서 해당 파일 삭제하고 impl로 되어 있는 것은 

class ExchangeService로 바꾸고 이거 하나만 쓸겁니다~

Jaejeong-Lee replied:
Service 하나로 하겠습니당

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+@Log4j2
+@Service
+@RequiredArgsConstructor
+public class ExchangeRateServiceImpl implements ExchangeRateService {
```
public class ExchangeRateService { ... 이런식으로 되는거죠

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+    /* 환율 API 호출을 위한 RestTemplate */
+    private final RestTemplate restTemplate = new RestTemplate();
+    /* Json 데이터를 Java 객체로 변환을 위한 ObjectMapper */
+    private final ObjectMapper objectMapper = new ObjectMapper();
```
p1 : 생성자 주입은 필요한 함수 안에서 하는게 맞을거 같아요 매번 해당 Service 가 호출될때 마다 해당 데이터를 쓰지도 않는데 생성자 주입을 할 필요는 없습니다. 

그리고 주석 삭제 부탁드려요   결론 아래와 같이 
```
private final ExchangeRateApiMapper apiMapper;
private final RestTemplate restTemplate;
private final ObjectMapper;
```

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+    public void fetchAndSaveExchangeRates() {
+
+        // 호출 시 날짜 정보 (yyyyMMdd 형식)
+        LocalDateTime exchangeRateDate = LocalDateTime.now();
```
p1 : 예성님 PR 한번 읽어봐주세요 뭔가 연관된 문제 생길거 같아요

hellozo0 replied:
mmaybe......

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+    private String API_KEY;
+
+
+    /* 환율 API로 환율 데이터 가져오는 함수 */
```
p1 : 함수에 대한 상세 설명을 할거면 해당 주석 스타일 보다 
```
/***
*
*
*/ 
``` 
이형식 안에 넣어주세요 그리고 /*** 치고 엔터 누르면 자동으로 형태 만들어 집니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+        String formattedDate = exchangeRateDate.toLocalDate().format(DateTimeFormatter.ofPattern("yyyyMMdd"));
+
+        // 환율 API
+        String url = "https://oapi.koreaexim.go.kr/site/program/financial/exchangeJSON?authkey=" + API_KEY + "&searchdate="+ formattedDate +"&data=AP01";
```
p1 : 그리고 이 부분은 상수와 같이 고정된 같이라서 해당 함수가 아니라 아예 밖에 final String url = ~ 이런식으로 빼는게 맞을거 같아요

Jaejeong-Lee replied:
넵 baseURL 부분으로 나눠서 최대한 밖으로 뺐습니다. 뒷부분은 변수를 삽입해야해서 함수 안에 뒀습니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
```
p1 : 이부분도 저같으면 dto로 매핑 하는 부분에 대한 함수를 따로 뺄거 같아요, 그리고 e.printStackTrace()만 할거면 불러오기 실패 했을때 어떤 Exception을 보여줄건지도 표현이 지금 안되어 있는거 같습니다! 

catch (Exception e) {
        log.error("환율 API 호출 및 저장 중 오류 발생", e);
        throw new RuntimeException("환율 데이터 처리 실패", e);
    }

와 같이 할 수 있을거 같고 "환율 API 호출 및 저장 중 오류 발생", "환율 데이터 처리 실패", 이런 에러 상수들은 
common > exception > enums > ErrorCode안에 정의하시고 불러와서 쓰시면 되세요

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+
+
+    /* 환율 API로 환율 데이터 가져오는 함수 */
+    @Override
```
p1 : interface 안 쓰니까 이것도 빼야합니다

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/dto/ExchangeRateApiDTO.java:
```diff
@@ -0,0 +1,30 @@
+@JsonIgnoreProperties(ignoreUnknown = true) // 필요없는 json 데이터 제외
+public class ExchangeRateApiDTO {
+
+    private Long exchangeRateId;
```
p1 : 이 부분도 우리가 Id 값을 저장하는게 아니라 나머지 값을 저장하면 DB에서 Id 값을 할당해주는거라 해당 컬럼은 삭제 되어야합니다.

Jaejeong-Lee replied:
해당 변수를 지워봤는데, mapper단에서 연결이 되어야 해서 그대로 유지하고자 합니다. 환율 저장할 때 자동으로 auto_increment로 들어가는 값이 여기에 해당해서 선언은 그대로 두어야 겠습니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/dto/ExchangeRateDTO.java:
```diff
@@ -0,0 +1,18 @@
+
+@Data
+public class ExchangeRateDTO {
+    private Long exchangeRateId;
```
p1 : 이 부분도 우리가 Id 값을 저장하는게 아니라 나머지 값을 저장하면 DB에서 Id 값을 할당해주는거라 해당 컬럼은 삭제 되어야합니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/service/exchange/ExchangeRateServiceImpl.java:
최종 결말 : 

아래와 같은 형식으로까지 극단적으로 함수를 쪼갤 수 있을거 같아요 
여기에서 너무 함수가 나눠진다고 생각하는 부분은 합치거나 해주세요! 

그리고 해당 로직들은 어떻게 보면 외부 api 끌어다가 쓰는거라서 우리 자체 내부 로직이라기 보다는 external 폴더에 좀더 적합한것 같습니다. 
ExchangeRateService는 유저가 어떤 동작을 하는 내부 비즈니스 로직 용도가 맞는거 같아요 

```
package org.scoula.service.exchange;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.core.type.TypeReference;

import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;

import org.scoula.controller.exchange.dto.ExchangeRateApiDTO;
import org.scoula.controller.exchange.dto.ExchangeRateDTO;
import org.scoula.mapper.exchange.ExchangeRateApiMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

@Log4j2
@Service
@RequiredArgsConstructor
public class ExchangeRateServiceImpl implements ExchangeRateService {

	private final ExchangeRateApiMapper apiMapper;
	private final RestTemplate restTemplate = new RestTemplate(); //이거 함수 안에서 주입하도록 수정해주세요
	private final ObjectMapper objectMapper = new ObjectMapper(); //이거 함수 안에서 주입하도록 수정해주세요

	@Value("${exchange.api.key}")
	private String API_KEY;

    @Override
    @Scheduled(cron = " 0 5 11 * * * ") // 11시 5분 스케쥴러 코드 실행
    @Transactional
    public void fetchAndSaveExchangeRates() {
        try {
            // 1. API URL 생성
            String url = buildApiUrl();

            // 2. API 호출 및 데이터 변환 후 저장
            fetchAndSaveFromApi(url);
        } catch (Exception e) {
            log.error("환율 데이터 가져오기 및 저장 중 오류 발생", e);
            //"환율 데이터 가져오기 및 저장 중 오류 발생 " ===> 이런거는 Exception 상수니까 ErrorCode안에 작성하고 ErrorCode.EXCHANGE_FAIL_ERROR 이런식으로 불러와주세요
        }
    }

    /**
     * API URL 생성
     */
    private String buildApiUrl() {
        String formattedDate = formatExchangeDate();
        return "https://oapi.koreaexim.go.kr/site/program/financial/exchangeJSON?authkey="
            + API_KEY + "&searchdate=" + formattedDate + "&data=AP01";
    }

    /**
     * API 호출 및 데이터 저장
     */
    private void fetchAndSaveFromApi(String url) throws Exception {
        // API 호출하여 JSON 데이터 가져오기
        String json = restTemplate.getForObject(url, String.class);

        // JSON을 DTO 리스트로 변환
        List<ExchangeRateApiDTO> apiDtoList = objectMapper.readValue(json,
            new TypeReference<List<ExchangeRateApiDTO>>() {});

        // Stream을 사용한 데이터 변환 및 저장
        saveExchangeRatesWithStream(apiDtoList);
    }

    /**
     * Stream + 생성자 패턴을 사용한 환율 데이터 변환 및 저장
     */
    private void saveExchangeRatesWithStream(List<ExchangeRateApiDTO> apiDtoList) {
        LocalDateTime now = LocalDateTime.now();
        LocalDateTime exchangeRateDate = calculateExchangeRateDate();

        apiDtoList.stream()
            .map(dto -> {
                // 환율 파싱 (콤마 제거 후 Double 변환)
                Double baseRate = parseExchangeRate(dto.getBaseExchangeRate());

                // 생성자를 사용한 객체 생성 (setter 대신)
                return new ExchangeRateDTO(
                    dto.getCurrencyCode(),           // 통화코드
                    dto.getCurrencyName(),           // 통화이름
                    baseRate,                        // 매매 기준율
                    baseRate,                        // 살 때 환율
                    baseRate,                        // 팔 때 환율
                    exchangeRateDate,                // 환율날짜
                    now,                             // 생성일시
                    now                              // 수정일시
                );
            })
            .forEach(entity -> apiMapper.insertDataToVO(entity)); // DB에 저장
    }

    /**
     * 환율 조회 기준 날짜 계산 (토요일/일요일은 금요일로 조정)
     */
    private LocalDateTime calculateExchangeRateDate() {
        LocalDateTime exchangeRateDate = LocalDateTime.now();
        int dayOfWeek = exchangeRateDate.getDayOfWeek().getValue();

        if (dayOfWeek == 6) { // 토요일
            exchangeRateDate = exchangeRateDate.minusDays(1);
        } else if (dayOfWeek == 7) { // 일요일
            exchangeRateDate = exchangeRateDate.minusDays(2);
        }

        return exchangeRateDate;
    }

    /**
     * 환율 날짜를 문자열로 포맷팅 (yyyyMMdd)
     */
    private String formatExchangeDate() {
        LocalDateTime exchangeRateDate = LocalDateTime.now();
        int dayOfWeek = exchangeRateDate.getDayOfWeek().getValue();

        if (dayOfWeek == 6) { // 토요일
            exchangeRateDate = exchangeRateDate.minusDays(1);
        } else if (dayOfWeek == 7) { // 일요일
            exchangeRateDate = exchangeRateDate.minusDays(2);
        }
        return exchangeRateDate
            .toLocalDate()
            .format(DateTimeFormatter.ofPattern("yyyyMMdd"));
    }

    /**
     * 환율 문자열을 Double로 변환 (콤마 제거)
     */
    private Double parseExchangeRate(String rawExchangeRate) {
        if (rawExchangeRate == null || rawExchangeRate.trim().isEmpty()) {
            throw new IllegalArgumentException("환율 데이터가 비어있습니다: " + rawExchangeRate);
            //"환율 데이터가 비어있습니다: " ===> 이런거는 Exception 상수니까 ErrorCode안에 작성하고 ErrorCode.EXCHANGE_EMPTY_ERROR 이런식으로 불러와주세요
        }

        String withoutComma = rawExchangeRate.replace(",", "");
        return Double.parseDouble(withoutComma);
    }
}

```

hellozo0 reviewed (commented):
으악 리뷰를 너무 중구 난방으로 한거 같아서 한번 할 수 있는 만큼만 수정해보시고 너무 어렵다 싶으면 불러주세요 내일 설명 드리면서 어떻게 해야하는지 같이 작성 도와드릴께요

Jaejeong-Lee committed:
[refactor] PR 피드백 수정
- external 폴더에 ExchangeRateApiService로 service 분리
- 필요없는 주석 삭제
- swagger 테스트용 어노테이션 추가
- 성공 응답 처리

Jaejeong-Lee committed:
[fix] git pull 충돌 처리

Jaejeong-Lee committed:
[chore] 파일 위치 변경

hellozo0 commented on trippy-server/exchange.sql:
p1 : 이거 이제 삭제 해주시고 메모에 넣어 두셨다고 merge 되면 디코 > sql 쌓는곳에 올려주세요!

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+
+@Api(tags = "Exchange")
+@Log4j2
```
p1 : Controller에서는 Logging을 하지 않고 Service에서만 할거라서 삭제 부탁드려요

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,61 @@
+
+    private final ExchangeRateService service;
+
+    /* 환율 정보 저장하는 요청 처리 */
```
p1 : 지금 보면 주석 스타일이 통일이 안되어 있어요 
그리고 Controller의 경우에는 스웨거 문서를 쓰다 보니 주석보다 @ApiOperation과 같은 걸로 충분하다고 생각합니다., 주석보다는 스웨거 문서 notes나 desciption을 더 추가로 작성하는게 좋을거 같습니다. 

그리고 Service에서 주석을 쓸때는 /*** 하고 엔터 누르시고 그 안에 작성하는걸로 통일하려고 합니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,61 @@
+@RequestMapping("/exchange-rate")
+public class ExchangeController {
+
+    private final ExchangeRateService service;
```
p2 : 그리고 대체로 service라고 변수명 짓기보다 exchangeRateService 라고 이름 짓는게 관례입니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,61 @@
+        return ResponseEntity.ok(response);
+    }
+
+}
```
p3 : 이거 나중에 작성하려고 만드신거 같은데 이번말고 나중에는 Issue에 해당하는 코드만 작성해주세요.

hellozo0 commented on trippy-server/src/main/java/org/scoula/controller/exchange/ExchangeController.java:
```diff
@@ -0,0 +1,61 @@
+    })
+    @PostMapping()
+    public SuccessNonDataResponse SaveExchangeRates() {
+        service.fetchAndSaveExchangeRates();
```
p1 : 그리고 코드 intelliJ에서 보면 해당 service안에 해당 함수 없습니다... 실행되는지 한번 더 테스트하고 Push 해주세요

hellozo0 replied:
exchangeRateAPIService 여기에 있는거 같아요 애를 final로 붙러와서 해야합니다.. 매칭이 잘못되어있어요

hellozo0 commented on trippy-server/src/main/java/org/scoula/domain/exchange/ExchangeRateVO.java:
```diff
@@ -20,8 +19,17 @@
+
+	@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
 	private LocalDateTime exchangeRateDate; //환율 고시날짜?
+
+	@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
+	private LocalDateTime createdAt;
+
+	@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
+	private LocalDateTime updatedAt;
 }
```
p1 : 지금 이부분은 extends BaseTime 해서 중복되는거 같아요 해당 부분 삭제 부탁드려요

Jaejeong-Lee replied:
형식을 저 상태로 저장하려고 createdAt, updatedAt이 중복되지만 적어뒀는데, 저장 형태를 좀 더 고민해봐야겠네요.

hellozo0 commented on trippy-server/src/main/java/org/scoula/domain/exchange/ExchangeRateVO.java:
```diff
@@ -20,8 +19,17 @@
+	private Double rateBuy;
+	private Double rateSell;
+
+	@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
```
just 기록 => 이걸 해두면 이 형식대로 우리가 보내준다~ 오케이

hellozo0 commented on trippy-server/src/main/java/org/scoula/external/exchangeApi/ExchangeRateAPIService.java:
이거 폴더링 명을 exchangeAPi 가 아니라 exchange로 해도 될거 같아요

hellozo0 replied:
애초에 external에 있다는게 외부 api를 쓴다는 의미입니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/mapper/exchange/ExchangeRateApiMapper.java:
```diff
@@ -0,0 +1,9 @@
+
+@Mapper
+public interface ExchangeRateApiMapper {
+    void saveDataToVO(ExchangeRateDTO exchangeRateDTO);
```
p1 : save 로만 해도 될거 같아요

hellozo0 commented on trippy-server/src/main/java/org/scoula/external/exchangeApi/ExchangeRateAPIService.java:
```diff
@@ -0,0 +1,88 @@
+
+            // API DTO -> 환율 DTO로 변환
+            for (ExchangeRateApiDTO dto : apiDtoList) {
+                ExchangeRateDTO entity = new ExchangeRateDTO();
```
p1 : 그리고 이것도 entity 라는 이름의 변수보다는 exchangeRateDTO 이런식으로 합니다...

hellozo0 commented on trippy-server/src/main/java/org/scoula/external/exchangeApi/ExchangeRateAPIService.java:
```diff
@@ -0,0 +1,88 @@
+                entity.setUpdatedAt(LocalDateTime.now());
+
+                // DB에 저장
+                apiMapper.saveDataToVO(entity);
```
p1 :그러면 이곳에서 save(exchangeRateDTO) 라고 하면 아 DTO를 저장하는구나 바로 알 수 있으니까요

hellozo0 commented on trippy-server/src/main/java/org/scoula/external/exchangeApi/ExchangeRateAPIService.java:
```diff
@@ -0,0 +1,88 @@
+@RequiredArgsConstructor
+public class ExchangeRateAPIService {
+
+    private final ExchangeRateApiMapper apiMapper;
```
p1 : 이또한  exchangeRateMapper 이런식으로 씁니다.

hellozo0 commented on trippy-server/src/main/java/org/scoula/external/exchangeApi/ExchangeRateAPIService.java:
```diff
@@ -0,0 +1,88 @@
+    @Transactional
+    public void fetchAndSaveExchangeRates() {
+
+        // 호출 시 날짜 정보 (yyyyMMdd 형식)
+        LocalDateTime exchangeRateDate = LocalDateTime.now();
+
+        // 토요일, 일요일은 금요일 환율로 설정
+        int dayOfWeek = exchangeRateDate.getDayOfWeek().getValue();
+        if (dayOfWeek == 6) {
+            exchangeRateDate = exchangeRateDate.minusDays(1);
+        } else if (dayOfWeek == 7) {
+            exchangeRateDate = exchangeRateDate.minusDays(2);
+        }
+        String formattedDate = exchangeRateDate.toLocalDate().format(DateTimeFormatter.ofPattern("yyyyMMdd"));
+
+        // 환율 API
+        String url = BASE_URL + API_KEY + "&searchdate="+ formattedDate +"&data=AP01";
```
p1 : 이부분은 함수 따로 빼는게 좋을거 같아요 날짜 포매팅은 데이터 가져오는데 부수적인 부분이지 핵심 로직이 아니니 분리하는게 좋을거 같습니다 ㅠ

hellozo0 reviewed (commented):
저번에 남겻던 부분까지 공통적인 부분들도 있는데 한번 보시고 모르겠으면 다시 디코로 질문 남겨주세요~

hellozo0 reviewed (commented):
..add

Jaejeong-Lee committed:
[refactor] Pr 피드백 반영사항
- sql 파일 삭제
- 형태 class -> record로 변경
- service 파일들 수정 및 함수 분리

Jaejeong-Lee committed:
[refactor] Vo 매핑함수 ApiDTO로 이동 및 service 수정

hellozo0 committed:
feat : 환전 환율 data type 변경

hellozo0 committed:
style : 코드 스타일 수정

hellozo0 reviewed (approved):
자잘자잘한 코드 포매팅만 제가 처리 했습니다 수고하셨어요!

hellozo0 committed:
feat : git pull
